Mage.Sets/src/mage/cards/a/AnrakyrTheTraveller.java

```diff
@@ -0,0 +1,198 @@
+            return false;
+        }
+        
+        Set<Card> cards = player.getHand().getCards(filter, game);
+        cards.addAll(player.getGraveyard().getCards(filter, game));
+        
+        Map<UUID, List<Card>> cardMap = new HashMap<>();
+        for (Card card : cards) {
+            List<Card> castableComponents = getCastableComponents(card, filter, source, player, game, false);
+            if (!castableComponents.isEmpty()) {
+                cardMap.put(card.getId(), castableComponents);
+            }
+        }      
+        Card cardToCast;
+        switch (cardMap.size()) {
+            case 0:
+                return false;
+            default:
+                Cards castableCards = new CardsImpl(cardMap.keySet());
+                TargetCard target = new TargetCard(0, 1, Zone.ALL, filter);
+                target.withNotTarget(true);
+                player.choose(Outcome.Benefit, castableCards, target, source, game);
+                cardToCast = castableCards.get(target.getFirstTarget(), game);
+        }  
+
+        if (cardToCast == null) {
+            return false;
+        }
+
+
+        List<Card> partsToCast = cardMap.get(cardToCast.getId());
+        String partsInfo = partsToCast
+                .stream()
+                .map(MageObject::getIdName)
+                .collect(Collectors.joining(" or "));
+        if (cardToCast == null
+                || partsToCast.size() < 1
+                || !player.chooseUse(
+                Outcome.PlayForFree, "Cast spell by paying life equal to its mana value rather than paying its mana cost (" + partsInfo + ")?", source, game
+        )) {
+            return false;
+        }
+        partsToCast.forEach(card -> game.getState().setValue("PlayFromNotOwnHandZone" + card.getId(), Boolean.TRUE));
+        
+        // pay life
+        PayLifeCost lifeCost = new PayLifeCost(cardToCast.getSpellAbility().getManaCosts().manaValue());
+        Costs<Cost> newCosts = new CostsImpl<>();
+        newCosts.add(lifeCost);
+        newCosts.addAll(cardToCast.getSpellAbility().getCosts());
+        player.setCastSourceIdWithAlternateMana(cardToCast.getId(), null, newCosts);
+
+        ActivatedAbility chosenAbility;
+        chosenAbility = player.chooseAbilityForCast(cardToCast, game, true);
+        boolean result = false;
+        if (chosenAbility instanceof SpellAbility) {
+            result = player.cast(
+                    (SpellAbility) chosenAbility,
+                    game, true, new ApprovingObject(source, game)
+            );
+        }
+        partsToCast.forEach(card -> game.getState().setValue("PlayFromNotOwnHandZone" + card.getId(), null));
+        if (player.isComputer() && !result) {
+            cards.remove(cardToCast);
+        }
+        return result;
+        
+
+        
+    }
+
+    private static List<Card> getCastableComponents(Card cardToCast, FilterCard filter, Ability source, Player player, Game game, boolean playLand) {
```

Comment:
Is that sensibly different from `CardUtil.getCastableComponents` ?

Reply:
No, but `CardUtil.getCastableComponents` is private, so we could not use it here directly.

Reply:
I would not mind making it public and use it there instead, that seems like code we'd want to not have duplicates around. @xenohedron thought on that?

Reply:
Yeah I think making it public is fine, and I'd prefer that to duplicating the code here. Ideally there should be a doc comment to explain its purpose/usage.

Reply:
We made the function public and added a doc comment in [8da98c2](https://github.com/magefree/mage/pull/11160/commits/8da98c2890170eb0977d3e724dde482c0ef4eb10)